File: objc/Tests/AirgapTests.m
//
//  AirgapTests.m
//  Airgap
//
//  Roundtrip tests for encoder and decoder
//

#import <Foundation/Foundation.h>
#import "AGEncoder.h"
#import "AGDecoder.h"

#define TEST_ASSERT(condition, message) \
    if (!(condition)) { \
        NSLog(@"❌ FAILED: %s - %@", #condition, message); \
        return NO; \
    } else { \
        NSLog(@"✅ PASSED: %s", #condition); \
    }

#define TEST_ASSERT_EQUAL(actual, expected, message) \
    if ((actual) != (expected)) { \
        NSLog(@"❌ FAILED: Expected %@ but got %@ - %@", @(expected), @(actual), message); \
        return NO; \
    } else { \
        NSLog(@"✅ PASSED: %@ == %@", @(actual), @(expected)); \
    }

BOOL testBasicRoundtrip(void) {
    NSLog(@"\n=== Testing Basic Roundtrip ===");

    // Create test data
    NSString *testString = @"Hello, Airgap!";
    NSData *originalData = [testString dataUsingEncoding:NSUTF8StringEncoding];

    // Create encoder
    NSError *error = nil;
    AGEncoder *encoder = [[AGEncoder alloc] initWithData:originalData chunkSize:100 error:&error];
    TEST_ASSERT(encoder != nil, ([NSString stringWithFormat:@"Failed to create encoder: %@", error]));
    TEST_ASSERT(error == nil, @"Error should be nil");

    // Check chunk count
    NSUInteger chunkCount = encoder.chunkCount;
    NSLog(@"Chunk count: %lu", (unsigned long)chunkCount);
    TEST_ASSERT(chunkCount > 0, @"Should have at least one chunk");

    // Get session ID
    uint32_t sessionId = encoder.sessionId;
    NSLog(@"Session ID: %u", sessionId);
    TEST_ASSERT(sessionId != 0, @"Session ID should not be zero");

    // Create decoder
    AGDecoder *decoder = [[AGDecoder alloc] init];
    TEST_ASSERT(decoder != nil, @"Failed to create decoder");

    // Process all chunks
    for (NSUInteger i = 0; i < chunkCount; i++) {
        // Get QR string (this is what would be scanned from a QR code)
        NSString *qrString = [encoder getQRStringAtIndex:i error:&error];
        TEST_ASSERT(qrString != nil, ([NSString stringWithFormat:@"Failed to get QR string at index %lu: %@", i, error]));
        TEST_ASSERT(qrString.length > 0, @"QR string should not be empty");

        // Process in decoder
        AGQRResult *qrResult = [decoder processQRString:qrString error:&error];
        TEST_ASSERT(qrResult != nil, ([NSString stringWithFormat:@"Failed to process QR at index %lu: %@", i, error]));
        TEST_ASSERT_EQUAL(qrResult.chunkNumber, i, @"Chunk number mismatch");
        TEST_ASSERT_EQUAL(qrResult.totalChunks, chunkCount, @"Total chunks mismatch");

        NSLog(@"Processed chunk %lu/%lu", (unsigned long)qrResult.chunkNumber + 1, (unsigned long)qrResult.totalChunks);
    }

    // Check decoder is complete
    TEST_ASSERT(decoder.isComplete, @"Decoder should be complete");
    TEST_ASSERT_EQUAL(decoder.receivedChunks, chunkCount, @"Received chunks mismatch");
    TEST_ASSERT_EQUAL(decoder.totalChunks, chunkCount, @"Total chunks mismatch");
    TEST_ASSERT_EQUAL((uint32_t)decoder.sessionId, sessionId, @"Session ID mismatch");

    // Get decoded data
    NSData *decodedData = [decoder getDataWithError:&error];
    TEST_ASSERT(decodedData != nil, ([NSString stringWithFormat:@"Failed to get decoded data: %@", error]));
    TEST_ASSERT_EQUAL(decodedData.length, originalData.length, @"Decoded data length mismatch");
    TEST_ASSERT([decodedData isEqualToData:originalData], @"Decoded data does not match original");

    NSString *decodedString = [[NSString alloc] initWithData:decodedData encoding:NSUTF8StringEncoding];
    NSLog(@"Original: %@", testString);
    NSLog(@"Decoded:  %@", decodedString);
    TEST_ASSERT([decodedString isEqualToString:testString], @"Decoded string does not match original");

    return YES;
}

BOOL testLargeDataRoundtrip(void) {
    NSLog(@"\n=== Testing Large Data Roundtrip ===");

    // Create larger test data (5KB)
    NSMutableData *largeData = [NSMutableData dataWithCapacity:5000];
    for (int i = 0; i < 5000; i++) {
        uint8_t byte = i % 256;
        [largeData appendBytes:&byte length:1];
    }

    // Create encoder with smaller chunk size to test multiple chunks
    NSError *error = nil;
    AGEncoder *encoder = [[AGEncoder alloc] initWithData:largeData chunkSize:500 error:&error];
    TEST_ASSERT(encoder != nil, ([NSString stringWithFormat:@"Failed to create encoder: %@", error]));

    NSUInteger chunkCount = encoder.chunkCount;
    NSLog(@"Large data chunk count: %lu", (unsigned long)chunkCount);
    TEST_ASSERT(chunkCount > 5, @"Should have multiple chunks");

    // Create decoder
    AGDecoder *decoder = [[AGDecoder alloc] init];
    TEST_ASSERT(decoder != nil, @"Failed to create decoder");

    // Process all chunks in order
    for (NSUInteger i = 0; i < chunkCount; i++) {
        NSString *qrString = [encoder getQRStringAtIndex:i error:&error];
        TEST_ASSERT(qrString != nil, ([NSString stringWithFormat:@"Failed to get QR string at %lu: %@", i, error]));

        AGQRResult *qrResult = [decoder processQRString:qrString error:&error];
        TEST_ASSERT(qrResult != nil, ([NSString stringWithFormat:@"Failed to process chunk %lu: %@", i, error]));
    }

    TEST_ASSERT(decoder.isComplete, @"Decoder should be complete");

    // Get and verify decoded data
    NSData *decodedData = [decoder getDataWithError:&error];
    TEST_ASSERT(decodedData != nil, @"Failed to get decoded data");
    TEST_ASSERT([decodedData isEqualToData:largeData], @"Decoded data does not match original");

    NSLog(@"Successfully roundtripped %lu bytes in %lu chunks",
          (unsigned long)largeData.length, (unsigned long)chunkCount);

    return YES;
}

BOOL testOutOfOrderChunks(void) {
    NSLog(@"\n=== Testing Out of Order Chunks ===");

    // Create test data
    NSString *testString = @"Out of order test data";
    NSData *originalData = [testString dataUsingEncoding:NSUTF8StringEncoding];

    NSError *error = nil;
    AGEncoder *encoder = [[AGEncoder alloc] initWithData:originalData chunkSize:50 error:&error];
    TEST_ASSERT(encoder != nil, @"Failed to create encoder");

    NSUInteger chunkCount = encoder.chunkCount;
    AGDecoder *decoder = [[AGDecoder alloc] init];

    // Collect all QR strings first
    NSMutableArray<NSString *> *qrStrings = [NSMutableArray array];
    for (NSUInteger i = 0; i < chunkCount; i++) {
        NSString *qrString = [encoder getQRStringAtIndex:i error:&error];
        TEST_ASSERT(qrString != nil, ([NSString stringWithFormat:@"Failed to get QR string at %lu: %@", i, error]));
        [qrStrings addObject:qrString];
    }

    // Process in reverse order
    for (NSInteger i = chunkCount - 1; i >= 0; i--) {
        AGQRResult *qrResult = [decoder processQRString:qrStrings[i] error:&error];
        TEST_ASSERT(qrResult != nil, ([NSString stringWithFormat:@"Failed to process chunk in reverse: %@", error]));
    }

    TEST_ASSERT(decoder.isComplete, @"Decoder should be complete after processing all chunks");

    NSData *decodedData = [decoder getDataWithError:&error];
    TEST_ASSERT([decodedData isEqualToData:originalData], @"Decoded data does not match when processed out of order");

    NSLog(@"Successfully processed %lu chunks out of order", (unsigned long)chunkCount);

    return YES;
}

BOOL testDecoderReset(void) {
    NSLog(@"\n=== Testing Decoder Reset ===");

    NSString *testString = @"Reset test";
    NSData *originalData = [testString dataUsingEncoding:NSUTF8StringEncoding];

    NSError *error = nil;
    AGEncoder *encoder = [[AGEncoder alloc] initWithData:originalData chunkSize:100 error:&error];
    AGDecoder *decoder = [[AGDecoder alloc] init];

    // Process first chunk
    NSString *qrString = [encoder getQRStringAtIndex:0 error:&error];
    [decoder processQRString:qrString error:&error];

    TEST_ASSERT(decoder.receivedChunks == 1, @"Should have received 1 chunk");

    // Reset decoder
    [decoder reset];

    TEST_ASSERT(decoder.receivedChunks == 0, @"Received chunks should be 0 after reset");
    TEST_ASSERT(decoder.totalChunks == 0, @"Total chunks should be 0 after reset");
    TEST_ASSERT(!decoder.isComplete, @"Decoder should not be complete after reset");
    TEST_ASSERT(decoder.sessionId == -1, @"Session ID should be -1 after reset");

    NSLog(@"Decoder successfully reset");

    return YES;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSLog(@"\n🧪 Running Airgap ObjC Tests\n");

        BOOL allPassed = YES;

        allPassed &= testBasicRoundtrip();
        allPassed &= testLargeDataRoundtrip();
        allPassed &= testOutOfOrderChunks();
        allPassed &= testDecoderReset();

        if (allPassed) {
            NSLog(@"\n✅ All tests passed!");
            return 0;
        } else {
            NSLog(@"\n❌ Some tests failed");
            return 1;
        }
    }
}

File: android/airgap/src/test/kotlin/app/rkz/airgap/AirgapTests.kt
package app.rkz.airgap

import kotlin.test.*

/**
 * Roundtrip tests for Airgap encoder and decoder
 */
class AirgapTests {

    @Test
    fun testBasicRoundtrip() {
        println("\n=== Testing Basic Roundtrip ===")

        // Create test data
        val testString = "Hello, Airgap!"
        val originalData = testString.toByteArray(Charsets.UTF_8)

        // Create encoder
        val encoder = AirgapEncoder(originalData, chunkSize = 100)
        assertNotNull(encoder, "Encoder should not be null")

        // Check chunk count
        val chunkCount = encoder.chunkCount
        println("Chunk count: $chunkCount")
        assertTrue(chunkCount > 0, "Should have at least one chunk")

        // Get session ID
        val sessionId = encoder.sessionId
        println("Session ID: $sessionId")
        assertTrue(sessionId != 0, "Session ID should not be zero")

        // Create decoder
        val decoder = AirgapDecoder()
        assertNotNull(decoder, "Decoder should not be null")

        // Process all chunks
        for (i in 0 until chunkCount) {
            // Get QR string (this is what would be scanned from a QR code)
            val qrString = encoder.getQRString(i)
            assertNotNull(qrString, "QR string at index $i should not be null")
            assertTrue(qrString.isNotEmpty(), "QR string should not be empty")

            // Process in decoder
            val qrResult = decoder.processQrString(qrString)
            assertNotNull(qrResult, "QRResult should not be null")
            assertEquals(i, qrResult.chunkNumber, "Chunk number mismatch")
            assertEquals(chunkCount, qrResult.totalChunks, "Total chunks mismatch")

            println("Processed chunk ${qrResult.chunkNumber + 1}/${qrResult.totalChunks}")
        }

        // Check decoder is complete
        assertTrue(decoder.isComplete, "Decoder should be complete")
        assertEquals(chunkCount, decoder.receivedChunks, "Received chunks mismatch")
        assertEquals(chunkCount, decoder.totalChunks, "Total chunks mismatch")
        assertEquals(sessionId, decoder.sessionId, "Session ID mismatch")

        // Get decoded data
        val decodedData = decoder.getData()
        assertNotNull(decodedData, "Decoded data should not be null")
        assertEquals(originalData.size, decodedData.size, "Decoded data length mismatch")
        assertContentEquals(originalData, decodedData, "Decoded data does not match original")

        val decodedString = decodedData.toString(Charsets.UTF_8)
        println("Original: $testString")
        println("Decoded:  $decodedString")
        assertEquals(testString, decodedString, "Decoded string does not match original")

        // Cleanup
        encoder.close()
        decoder.close()
    }

    @Test
    fun testLargeDataRoundtrip() {
        println("\n=== Testing Large Data Roundtrip ===")

        // Create larger test data (5KB)
        val largeData = ByteArray(5000) { (it % 256).toByte() }

        // Create encoder with smaller chunk size to test multiple chunks
        val encoder = AirgapEncoder(largeData, chunkSize = 500)
        assertNotNull(encoder, "Encoder should not be null")

        val chunkCount = encoder.chunkCount
        println("Large data chunk count: $chunkCount")
        assertTrue(chunkCount > 5, "Should have multiple chunks")

        // Create decoder
        val decoder = AirgapDecoder()
        assertNotNull(decoder, "Decoder should not be null")

        // Process all chunks in order
        for (i in 0 until chunkCount) {
            val qrString = encoder.getQRString(i)
            assertNotNull(qrString, "QR string at index $i should not be null")

            val qrResult = decoder.processQrString(qrString)
            assertNotNull(qrResult, "QRResult at index $i should not be null")
        }

        assertTrue(decoder.isComplete, "Decoder should be complete")

        // Get and verify decoded data
        val decodedData = decoder.getData()
        assertNotNull(decodedData, "Decoded data should not be null")
        assertContentEquals(largeData, decodedData, "Decoded data does not match original")

        println("Successfully roundtripped ${largeData.size} bytes in $chunkCount chunks")

        // Cleanup
        encoder.close()
        decoder.close()
    }

    @Test
    fun testOutOfOrderChunks() {
        println("\n=== Testing Out of Order Chunks ===")

        // Create test data
        val testString = "Out of order test data"
        val originalData = testString.toByteArray(Charsets.UTF_8)

        val encoder = AirgapEncoder(originalData, chunkSize = 50)
        assertNotNull(encoder, "Encoder should not be null")

        val chunkCount = encoder.chunkCount
        val decoder = AirgapDecoder()

        // Collect all QR strings first
        val qrStrings = mutableListOf<String>()
        for (i in 0 until chunkCount) {
            val qrString = encoder.getQRString(i)
            assertNotNull(qrString, "QR string at index $i should not be null")
            qrStrings.add(qrString)
        }

        // Process in reverse order
        for (i in chunkCount - 1 downTo 0) {
            val qrResult = decoder.processQrString(qrStrings[i])
            assertNotNull(qrResult, "QRResult should not be null when processing in reverse")
        }

        assertTrue(decoder.isComplete, "Decoder should be complete after processing all chunks")

        val decodedData = decoder.getData()
        assertContentEquals(originalData, decodedData, "Decoded data does not match when processed out of order")

        println("Successfully processed $chunkCount chunks out of order")

        // Cleanup
        encoder.close()
        decoder.close()
    }

    @Test
    fun testDecoderReset() {
        println("\n=== Testing Decoder Reset ===")

        val testString = "Reset test"
        val originalData = testString.toByteArray(Charsets.UTF_8)

        val encoder = AirgapEncoder(originalData, chunkSize = 100)
        val decoder = AirgapDecoder()

        // Process first chunk
        val qrString = encoder.getQRString(0)
        decoder.processQrString(qrString)

        assertEquals(1, decoder.receivedChunks, "Should have received 1 chunk")

        // Reset decoder
        decoder.reset()

        assertEquals(0, decoder.receivedChunks, "Received chunks should be 0 after reset")
        assertEquals(0, decoder.totalChunks, "Total chunks should be 0 after reset")
        assertFalse(decoder.isComplete, "Decoder should not be complete after reset")
        assertEquals(-1, decoder.sessionId, "Session ID should be -1 after reset")

        println("Decoder successfully reset")

        // Cleanup
        encoder.close()
        decoder.close()
    }

    @Test
    fun testMultipleEncoders() {
        println("\n=== Testing Multiple Encoders ===")

        val data1 = "First encoder data".toByteArray(Charsets.UTF_8)
        val data2 = "Second encoder data".toByteArray(Charsets.UTF_8)

        val encoder1 = AirgapEncoder(data1, chunkSize = 100)
        val encoder2 = AirgapEncoder(data2, chunkSize = 100)

        // Session IDs should be different
        assertNotEquals(encoder1.sessionId, encoder2.sessionId, "Session IDs should be different")

        val decoder1 = AirgapDecoder()
        val decoder2 = AirgapDecoder()

        // Decode data from encoder1
        for (i in 0 until encoder1.chunkCount) {
            decoder1.processQrString(encoder1.getQRString(i))
        }

        // Decode data from encoder2
        for (i in 0 until encoder2.chunkCount) {
            decoder2.processQrString(encoder2.getQRString(i))
        }

        assertTrue(decoder1.isComplete, "Decoder 1 should be complete")
        assertTrue(decoder2.isComplete, "Decoder 2 should be complete")

        val decoded1 = decoder1.getData()
        val decoded2 = decoder2.getData()

        assertContentEquals(data1, decoded1, "Decoded data 1 should match")
        assertContentEquals(data2, decoded2, "Decoded data 2 should match")

        println("Successfully handled multiple independent encoder/decoder pairs")

        // Cleanup
        encoder1.close()
        encoder2.close()
        decoder1.close()
        decoder2.close()
    }

    @Test
    fun testEmptyDataThrowsException() {
        println("\n=== Testing Empty Data Throws Exception ===")

        assertFailsWith<AirgapException>("Should throw exception for empty data") {
            AirgapEncoder(ByteArray(0), chunkSize = 100)
        }

        println("Empty data correctly throws exception")
    }

    @Test
    fun testInvalidChunkSizeThrowsException() {
        println("\n=== Testing Invalid Chunk Size Throws Exception ===")

        val testData = "Test".toByteArray(Charsets.UTF_8)

        assertFailsWith<AirgapException>("Should throw exception for chunk size too small") {
            AirgapEncoder(testData, chunkSize = 10)
        }

        assertFailsWith<AirgapException>("Should throw exception for chunk size too large") {
            AirgapEncoder(testData, chunkSize = 2000)
        }

        println("Invalid chunk sizes correctly throw exceptions")
    }
}

fun main() {
    println("\n🧪 Running Airgap Kotlin Tests\n")

    val tests = AirgapTests()
    var allPassed = true

    val testMethods = listOf(
        "testBasicRoundtrip" to { tests.testBasicRoundtrip() },
        "testLargeDataRoundtrip" to { tests.testLargeDataRoundtrip() },
        "testOutOfOrderChunks" to { tests.testOutOfOrderChunks() },
        "testDecoderReset" to { tests.testDecoderReset() },
        "testMultipleEncoders" to { tests.testMultipleEncoders() },
        "testEmptyDataThrowsException" to { tests.testEmptyDataThrowsException() },
        "testInvalidChunkSizeThrowsException" to { tests.testInvalidChunkSizeThrowsException() }
    )

    for ((name, test) in testMethods) {
        try {
            test()
            println("✅ PASSED: $name\n")
        } catch (e: Exception) {
            println("❌ FAILED: $name")
            println("   Error: ${e.message}\n")
            e.printStackTrace()
            allPassed = false
        }
    }

    if (allPassed) {
        println("\n✅ All tests passed!")
        System.exit(0)
    } else {
        println("\n❌ Some tests failed")
        System.exit(1)
    }
}